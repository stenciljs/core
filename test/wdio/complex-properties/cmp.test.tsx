/// <reference types="../dist/components.d.ts" />
import { render } from '@wdio/browser-runner/stencil';
import { $, expect } from '@wdio/globals';

import { renderToString, serializeProperty } from '../hydrate/index.mjs';

const template = `<complex-properties
  foo=${serializeProperty({ bar: 123, loo: [1, 2, 3], qux: { quux: Symbol('quux') } })}
  baz=${serializeProperty(new Map([['foo', { qux: Symbol('quux') }]]))}
  quux=${serializeProperty(new Set(['foo']))}
  corge=${serializeProperty(new Set([{ foo: { bar: 'foo' } }]))}
  grault=${serializeProperty(Infinity)}
  waldo=${serializeProperty(null)}
  kids-names=${serializeProperty(['John', 'Jane', 'Jim'])}
/>`;

describe('complex-properties', () => {
  it('should render complex properties', async () => {
    const { html } = await renderToString(template, {
      prettyHtml: true,
      fullDocument: false,
    });
    expect(html).toMatchSnapshot();
  });

  it('can render component and update properties', async () => {
    const { html } = await renderToString(template, {
      fullDocument: false,
    });

    render({ html, components: [] });
    await expect($('complex-properties')).toHaveText(
      [
        `this.foo.bar: 123`,
        `this.foo.loo: 1, 2, 3`,
        `this.foo.qux: symbol`,
        `this.baz.get('foo'): symbol`,
        `this.quux.has('foo'): true`,
        `this.grault: true`,
        `this.waldo: true`,
        `this.kidsNames: John, Jane, Jim`,
      ].join('\n'),
    );
  });

  it('can change a complex property and see it updated correctly', async () => {
    const elm = document.querySelector('complex-properties') as HTMLComplexPropertiesElement;
    elm.foo = { bar: '456', loo: [4, 5, 6], qux: { quux: Symbol('new quux') } };
    elm.kidsNames.push('Jill');
    await expect(elm).toHaveText(
      [
        `this.foo.bar: 456`,
        `this.foo.loo: 4, 5, 6`,
        `this.foo.qux: symbol`,
        `this.baz.get('foo'): symbol`,
        `this.quux.has('foo'): true`,
        `this.grault: true`,
        `this.waldo: true`,
        `this.kidsNames: John, Jane, Jim, Jill`,
      ].join('\n'),
    );
  });
});
