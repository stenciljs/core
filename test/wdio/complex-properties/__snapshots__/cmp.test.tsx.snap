// Snapshot v1

exports[`complex-properties > should render complex properties 1`] = `
"<complex-properties baz="serialized:eyJ0eXBlIjoibWFwIiwidmFsdWUiOltbeyJ0eXBlIjoic3RyaW5nIiwidmFsdWUiOiJmb28ifSx7InR5cGUiOiJvYmplY3QiLCJ2YWx1ZSI6W1sicXV4Iix7InR5cGUiOiJzeW1ib2wiLCJ2YWx1ZSI6InF1dXgifV1dfV1dfQ==" class="hydrated sc-complex-properties-h" corge="serialized:eyJ0eXBlIjoic2V0IiwidmFsdWUiOlt7InR5cGUiOiJvYmplY3QiLCJ2YWx1ZSI6W1siZm9vIix7InR5cGUiOiJvYmplY3QiLCJ2YWx1ZSI6W1siYmFyIix7InR5cGUiOiJzdHJpbmciLCJ2YWx1ZSI6ImZvbyJ9XV19XV19XX0=" foo="serialized:eyJ0eXBlIjoib2JqZWN0IiwidmFsdWUiOltbImJhciIseyJ0eXBlIjoibnVtYmVyIiwidmFsdWUiOjEyM31dLFsibG9vIix7InR5cGUiOiJhcnJheSIsInZhbHVlIjpbeyJ0eXBlIjoibnVtYmVyIiwidmFsdWUiOjF9LHsidHlwZSI6Im51bWJlciIsInZhbHVlIjoyfSx7InR5cGUiOiJudW1iZXIiLCJ2YWx1ZSI6M31dfV0sWyJxdXgiLHsidHlwZSI6Im9iamVjdCIsInZhbHVlIjpbWyJxdXV4Iix7InR5cGUiOiJzeW1ib2wiLCJ2YWx1ZSI6InF1dXgifV1dfV1dfQ==" grault="serialized:eyJ0eXBlIjoibnVtYmVyIiwidmFsdWUiOiJJbmZpbml0eSJ9" kids-names="serialized:eyJ0eXBlIjoiYXJyYXkiLCJ2YWx1ZSI6W3sidHlwZSI6InN0cmluZyIsInZhbHVlIjoiSm9obiJ9LHsidHlwZSI6InN0cmluZyIsInZhbHVlIjoiSmFuZSJ9LHsidHlwZSI6InN0cmluZyIsInZhbHVlIjoiSmltIn1dfQ==" quux="serialized:eyJ0eXBlIjoic2V0IiwidmFsdWUiOlt7InR5cGUiOiJzdHJpbmciLCJ2YWx1ZSI6ImZvbyJ9XX0=" s-id="1" waldo="serialized:eyJ0eXBlIjoibnVsbCJ9">
  <template shadowrootmode="open">
    <ul c-id="1.0.0.0" class="sc-complex-properties">
      <li c-id="1.1.1.0" class="sc-complex-properties">
        <!--t.1.2.2.0-->
        this.foo.bar: 123
      </li>
      <li c-id="1.3.1.1" class="sc-complex-properties">
        <!--t.1.4.2.0-->
        this.foo.loo: 1, 2, 3
      </li>
      <li c-id="1.5.1.2" class="sc-complex-properties">
        <!--t.1.6.2.0-->
        this.foo.qux: symbol
      </li>
      <li c-id="1.7.1.3" class="sc-complex-properties">
        <!--t.1.8.2.0-->
        this.baz.get('foo'): symbol
      </li>
      <li c-id="1.9.1.4" class="sc-complex-properties">
        <!--t.1.10.2.0-->
        this.quux.has('foo'): true
      </li>
      <li c-id="1.11.1.5" class="sc-complex-properties">
        <!--t.1.12.2.0-->
        this.grault: true
      </li>
      <li c-id="1.13.1.6" class="sc-complex-properties">
        <!--t.1.14.2.0-->
        this.waldo: true
      </li>
      <li c-id="1.15.1.7" class="sc-complex-properties">
        <!--t.1.16.2.0-->
        this.kidsNames: John, Jane, Jim
      </li>
    </ul>
  </template>
  <!--r.1-->
</complex-properties>"
`;
